src/ui/public/vislib/lib/axis.js

```diff
@@ -5,145 +5,113 @@ import ErrorHandlerProvider from 'ui/vislib/lib/_error_handler';
       super();
-      if (args.type === 'category') {
-        _.extend(this, defaults, categoryDefaults, args);
-      } else {
-        _.extend(this, defaults, args);
+      this.data = args.data;
+      this.config = new AxisConfig(args);
+      if (this.config.get('type') === 'category') {
+        this.values = this.data.xValues();
+        this.ordered = this.data.get('ordered');
       }
-
-      this._attr = args.vis._attr;
-      this.elSelector = this.elSelector.replace('{pos}', this.position);
-      this.scale = new AxisScale(this, {scale: this.scale});
-      this.axisTitle = new AxisTitle(this, this.axisTitle);
-      this.axisLabels = new AxisLabels(this, this.labels);
+      this.axisScale = new AxisScale(this.config, this.data);
+      this.axisTitle = new AxisTitle(this.config);
+      this.axisLabels = new AxisLabels(this.config, this.axisScale);
     }
 
     render() {
-      d3.select(this.vis.el).selectAll(this.elSelector).call(this.draw());
-    }
-
-    isHorizontal() {
-      return (this.position === 'top' || this.position === 'bottom');
+      const elSelector = this.config.get('elSelector');
+      const rootEl = this.config.get('rootEl');
+      d3.select(rootEl).selectAll(elSelector).call(this.draw());
     }
 
     getAxis(length) {
-      const scale = this.scale.getScale(length);
-
+      const scale = this.axisScale.getScale(length);
+      const position = this.config.get('position');
       return d3.svg.axis()
-        .scale(scale)
-        .tickFormat(this.tickFormat(this.domain))
-        .ticks(this.tickScale(length))
-        .orient(this.position);
+      .scale(scale)
+      .tickFormat(this.tickFormat(this.domain))
+      .ticks(this.tickScale(length))
+      .orient(position);
     }
 
     getScale() {
-      return this.scale.scale;
+      return this.axisScale.scale;
     }
 
     addInterval(interval) {
-      return this.scale.addInterval(interval);
+      return this.axisScale.addInterval(interval);
     }
 
     substractInterval(interval) {
-      return this.scale.substractInterval(interval);
+      return this.axisScale.substractInterval(interval);
     }
 
     tickScale(length) {
       const yTickScale = d3.scale.linear()
-        .clamp(true)
-        .domain([20, 40, 1000])
-        .range([0, 3, 11]);
+      .clamp(true)
+      .domain([20, 40, 1000])
+      .range([0, 3, 11]);
 
       return Math.ceil(yTickScale(length));
     }
 
     tickFormat() {
-      if (this.axisFormatter) return this.axisFormatter;
-      if (this.isPercentage()) return d3.format('%');
+      if (this.config.get('labels.axisFormatter')) return this.config.get('labels.axisFormatter');
+      if (this.config.isPercentage()) return d3.format('%');
       return d3.format('n');
     }
 
     getLength(el, n) {
-      if (this.isHorizontal()) {
-        return $(el).parent().width() / n - this._attr.margin.left - this._attr.margin.right - 50;
+      const margin = this.config.get('vis._attr.margin');
```

Comment:
its accessing vis config .... in the future we'll probably extract vis config in similar manner we did with axis config ?